Userland/Utilities/init.cpp

```diff
@@ -127,26 +137,65 @@ static ErrorOr<void> reopen_base_file_descriptors()
     // This affects also every other process that inherits the file descriptors
     // from SystemServer, so it is important for other things (also for ProcFS
     // tests that are running in CI mode).
-    int stdin_new_fd = TRY(Core::System::open("/dev/null"sv, O_NONBLOCK));
+    int stdin_new_fd = TRY(acquire_new_stdin_fd(emergency));
     TRY(Core::System::dup2(stdin_new_fd, 0));
     TRY(Core::System::dup2(stdin_new_fd, 1));
     TRY(Core::System::dup2(stdin_new_fd, 2));
     return {};
 }
 
-ErrorOr<int> serenity_main(Main::Arguments)
+static ErrorOr<void> execute_emergency_shell()
+{
+    outln("Emergency mode: Dropping to emergency shell mode");
+    outln("You may use this shell as rescue environment now.");
+    if (!Core::System::access("/bin/Shell"sv, X_OK, 0).is_error())
+        TRY(Core::System::exec("/bin/Shell"sv, Vector { "/bin/Shell"sv }, Core::System::SearchInPath::No));
+    if (!Core::System::access("/bin/BuggieBox"sv, X_OK, 0).is_error())
+        TRY(Core::System::exec("/bin/BuggieBox"sv, Vector { "/bin/BuggieBox"sv, "/bin/Shell"sv }, Core::System::SearchInPath::No));
+    outln("Failed to find a program to be used as rescue environment. Halting.");
+    VERIFY_NOT_REACHED();
+}
+
+ErrorOr<int> serenity_main(Main::Arguments arguments)
 {
     if (geteuid() != 0) {
         warnln("Not running as root :^(");
         return 1;
     }
-    TRY(mount_all_filesystems());
+
+    bool emergency = false;
+    // NOTE: What determins this flag normally is the user running the OS
+    // with a kernel commandline including "init_args=emergency".
+    if (arguments.strings.size() > 1 && arguments.strings[1] == "emergency"sv)
+        emergency = true;
+
+    // NOTE: The reason we check for emergency state is because we should avoid trying to mount
+    // anything if the user requested to use "emergency mode".
+    if (!emergency) {
+        // NOTE: If we are not in emergency state, try to mount filesystems according
+        // to the /etc/fstab file. If it fails, declare emergency state and drop to shell.
+        if (auto result = mount_all_filesystems(); result.is_error())
+            emergency = true;
+    }
+
+    // NOTE: If we are not in emergency state, and the /bin/SystemServer program is not accessible
+    // (or can't be run due to bad permissions) then declare emergency state and drop to shell.
+    // NOTE: The reason we check for emergency state is because we should avoid useless syscalls at this stage.
```

Comment:
I feel like these last three comments can just be normal comments, they aren't particularly noteworthy in terms of explaining non-obvious behavior.